grant-project/grant-monorepo
# User Authentication UI (Pt 1 - The Basics)

### What This Does

Begins laying the groundwork for auth. This implements an auth page, with a basic flow for web3 or entering your own address. It lays down some of the Redux state and actions, but they're just stubbed out. The PR was getting hefty, so I wanted to submit it bit by bit. In short, this does the following:

* Reworks the header to have auth state in the top right. Header has also been trimmed down a little bit.
    * Header auto detects web3 state and shows identicon of logged in user
* Adds a page at `/auth` and at `/profile`
    * `/profile` is just a stubbed, and will be replaced with whatever @AMStrix does for #120.
* Implements a faux auth flow for web3 and custom addresses
    * No message signing or actual `user` creation happens here, all in local memory
    * Hardware wallet buttons are there, but don't do anything yet
* Adds an `AuthRoute` component that forces you to /auth if signed out, or to /profile if you're sign in and it has the `onlyLoggedOut` prop set. Shows a loader while determining auth state.

### Things to Improve

* Implement message signing, even if only for frontend to better test flow
* Persist some auth state to redux store, and bootstrap auth state on page load
* Make signup actually create a new user
* Lookup if user exists on backend when address is provided, show sign in instead
* Implement hardware wallets
* Validate forms
* Add address provider icons to auth provider selection

### Steps to Test

* Load a page with an unlocked web3 browser, confirm that you see your identicon in top right
* Click on sign in, confirm you're immediately taken to web3 auth
* Complete signup, confirm you're taken to `/profile` with the correct info
* Load a page in a non-web3 browser, confirm you see no identicon in top right
* Click sign in, confirm you're taken to options of identity providers
* Select address and paste in an address
* Complete signup, confirm you're taken to `/profile` with the correct info

## Screenshots

### Auth providers

<img width="1034" alt="screen shot 2018-10-01 at 12 49 45 pm" src="https://user-images.githubusercontent.com/649992/46302962-8d2bb780-c578-11e8-9957-6836b6dbc530.png">

### Signup form

<img width="1036" alt="screen shot 2018-10-01 at 12 50 25 pm" src="https://user-images.githubusercontent.com/649992/46303013-b2202a80-c578-11e8-8d67-a831f767edcd.png">
